core/corehttp/hostname.go

```diff
@@ -0,0 +1,301 @@
+package corehttp
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"regexp"
+	"strings"
+
+	cid "github.com/ipfs/go-cid"
+	core "github.com/ipfs/go-ipfs/core"
+	namesys "github.com/ipfs/go-ipfs/namesys"
+
+	config "github.com/ipfs/go-ipfs-config"
+	nsopts "github.com/ipfs/interface-go-ipfs-core/options/namesys"
+	isd "github.com/jbenet/go-is-domain"
+)
+
+var pathGatewaySpec = config.GatewaySpec{
+	Paths:         []string{"/ipfs/", "/ipns/", "/api/", "/p2p/", "/version"},
+	UseSubdomains: false,
+}
+
+var subdomainGatewaySpec = config.GatewaySpec{
+	Paths:         []string{"/ipfs/", "/ipns/", "/api/", "/p2p/"},
+	UseSubdomains: true,
+}
+
+var defaultKnownGateways = map[string]config.GatewaySpec{
+	// Note: local IPs act as catch-all, no need to define them here
+	//"127.0.0.1":       pathGatewaySpec,
+	//"::1":             pathGatewaySpec,
+	"localhost":       subdomainGatewaySpec,
+	"ipfs.io":         pathGatewaySpec,
+	"gateway.ipfs.io": pathGatewaySpec,
+	"dweb.link":       subdomainGatewaySpec,
+}
+
+// Find content identifier, protocol, and remaining hostname (host+optional port)
+// of a subdomain gateway (eg. *.ipfs.foo.bar.co.uk)
+var subdomainGatewayRegex = regexp.MustCompile("^(.+).(ipfs|ipns|ipld|p2p|api).([^/?#&]+)$")
```

Comment:
Ack, this is a very good point. We should make it easier for people, not harder.

If someone whitelisted `/api` by adding it to `Paths` for `example.com`, they expect it to "just work", and that means no redirect, and supporting both `example.com/api/` (interop with path gateways and existing tools) and   `$cid.ipfs.example.com/api/` to avoid CORS issues when accessing API from JS loaded from subdomain content root Origin at `$cid.ipfs.example.com`.

I made that change in aecfa66f0561d1c915c014429689d1df7aa9751a (just `/api`,  `/p2p` was fixed in 835a8f06f977c7f4feb21dfa75c2668772617e05)

Reply:
> supporting both example.com/api/ (interop with path gateways and existing tools) and $cid.ipfs.example.com/api/ to avoid CORS issues when accessing API from JS loaded from subdomain content root Origin at $cid.ipfs.example.com.

I would _not_ support `$cid.ipfs.example.com/api` as that will conflict with paths within the `$cid.ipfs.example.com` directory. Is CORS really an issue here?

Reply:
Fair point. I'd rather have people setup CORS if they need to, than block them from using specific words in URLs on their websites. Removing path conflicts is more important.

I removed support for `$cid.ipfs.example.com/api` in dd8c804cd24cf844c68a7423570a34566ac384f7 and added bunch of tests to ensure `api` and `ipfs`  can be used as file or directory names in content roots.